### PR TITLE
lightning: sample once parquet file

### LINKED ISSUE #56104: lightning might stuck for hours when importing parquet files from cloud storage
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

after https://github.com/pingcap/tidb/pull/46984, including `>= v7.5.x`, `7.1.3+`, we will sample parquet all the time in serial, it's very slow and might takes hours if user have large mount of parquet files before lightning start doing import, and the time takes to sample the files might even longer than real import work.

we only need this size for displaying progress more accurately and use it as a reference when splitting engine, but slowing import this much is un-acceptable.

### 2. What did you expect to see? (Required)
start import fast
### 3. What did you see instead (Required)
it might takes hours before start doing any work

### PR DESCRIPTION
**Pull request by zeminzhou**
### What problem does this PR solve?

Issue Number: close #56104

Problem Summary:

### What changed and how does it work?
From test, before this pr, lightning importing ten files in parquet format needs to takes 4m40s in the preparation stage. After this pr, lightning importing ten files in parquet format needs to takes 1m10s.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/lightning/mydump/loader.go [hunk 1]
```diff
@@ -532,12 +536,24 @@ func (s *mdLoaderSetup) constructFileInfo(ctx context.Context, path string, size
...
 		s.tableDatas = append(s.tableDatas, info)
 	case SourceTypeParquet:
-		parquestDataSize, err2 := SampleParquetDataSize(ctx, info.FileMeta, s.loader.GetStore())
-		if err2 != nil {
-			logger.Error("fail to sample parquet data size", zap.String("category", "loader"),
-				zap.String("schema", res.Schema), zap.String("table", res.Name), zap.Stringer("type", res.Type), zap.Error(err2))
-		} else {
-			info.FileMeta.RealSize = parquestDataSize
+		tableName := info.TableName.String()
+		if s.sampledParquetRowSizes[tableName] == 0 {
```
Comment (lance6716) [inline pkg/lightning/mydump/loader.go:540]: can we use _, ok := map[...] to check? I'm afraid that there will be some zero row files, so `SampleParquetRowSize` will be called unnecessary.

File: pkg/lightning/mydump/loader.go [hunk 2]
```diff
@@ -823,8 +839,8 @@ func SampleFileCompressRatio(ctx context.Context, fileMeta SourceFileMeta, store
 	return float64(tot) / float64(pos), nil
 }
 
-// SampleParquetDataSize samples the data size of the parquet file.
-func SampleParquetDataSize(ctx context.Context, fileMeta SourceFileMeta, store storage.ExternalStorage) (int64, error) {
+// SampleParquetRowSize samples one row size of the parquet file.
```
Comment (lance6716) [inline pkg/lightning/mydump/loader.go:842]: what do you mean by "one row size"? I think row size is enough

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: pkg/lightning/mydump/loader.go (hunk 1 above)
Comment (zeminzhou) [reply pkg/lightning/mydump/loader.go:540]: If the samped files have zero row, we should continue to sample until the file has rows.

File: pkg/lightning/mydump/loader.go (hunk 2 above)
Comment (zeminzhou) [reply pkg/lightning/mydump/loader.go:842]: should be "row size", thanks!

File: pkg/lightning/mydump/loader.go [hunk 3]
```diff
@@ -532,12 +536,24 @@ func (s *mdLoaderSetup) constructFileInfo(ctx context.Context, path string, size
...
+		if s.sampledParquetRowSizes[tableName] == 0 {
+			s.sampledParquetRowSizes[tableName], err = SampleParquetRowSize(ctx, info.FileMeta, s.loader.GetStore())
+			if err != nil {
+				logger.Error("fail to sample parquet row size", zap.String("category", "loader"),
+					zap.String("schema", res.Schema), zap.String("table", res.Name),
+					zap.Stringer("type", res.Type), zap.Error(err))
+			}
+		}
+		if s.sampledParquetRowSizes[tableName] != 0 {
+			totalRowCount, err2 := ReadParquetFileRowCountByFile(ctx, s.loader.GetStore(), info.FileMeta)
```
Comment (D3Hunter) [inline pkg/lightning/mydump/loader.go:549]: this function runs for each file and in serial. and we will do it again in `importTables`, have you tested how long this takes?

for parquet, RealSize is only used in precheck/estimate-compaction-threshold/progress-related, so maybe revert https://github.com/pingcap/tidb/pull/46984 completely or disable it on default

Comment (zeminzhou) [reply pkg/lightning/mydump/loader.go:549]: In this pr, the `ReadParquetFileRowCountByFile` only run once for each table.

Comment (zeminzhou) [reply pkg/lightning/mydump/loader.go:549]: With this PR, `ReadParquetFileRowCountByFile` will no longer affect import time.

Comment (D3Hunter) [reply pkg/lightning/mydump/loader.go:549]: `SampleParquetRowSize` runs once if sampledParquetRowSizes is 0 for the table, it's not zero later. So for later files of the same table `ReadParquetFileRowCountByFile` is called

Comment (zeminzhou) [reply pkg/lightning/mydump/loader.go:549]: Sorry, I misunderstood `ReadParquetFileRowCountByFile` at first

Comment (zeminzhou) [reply pkg/lightning/mydump/loader.go:549]: It looks like we need to cache the row count for parquet

Comment (zeminzhou) [reply pkg/lightning/mydump/loader.go]: PTAL～

File: pkg/lightning/mydump/loader.go [hunk 4]
```diff
@@ -532,12 +536,27 @@ func (s *mdLoaderSetup) constructFileInfo(ctx context.Context, path string, size
...
+			s.sampledParquetRowSizes[tableName], err = SampleParquetRowSize(ctx, info.FileMeta, s.loader.GetStore())
+			if err != nil {
+				logger.Error("fail to sample parquet row size", zap.String("category", "loader"),
+					zap.String("schema", res.Schema), zap.String("table", res.Name),
+					zap.Stringer("type", res.Type), zap.Error(err))
+				return errors.Trace(err)
+			}
+		}
+		if s.sampledParquetRowSizes[tableName] != 0 {
+			totalRowCount, err := ReadParquetFileRowCountByFile(ctx, s.loader.GetStore(), info.FileMeta)
```
Comment (D3Hunter) [inline pkg/lightning/mydump/loader.go:550]: this part runs in serial, can you do some test about the ratio of time takes in this part

File: lightning/pkg/importer/import.go [hunk 5]
```diff
@@ -1488,19 +1488,8 @@ func (rc *Controller) importTables(ctx context.Context) (finalErr error) {
...
-				for i, fi := range tableMeta.DataFiles {
+				for _, fi := range tableMeta.DataFiles {
 					totalDataSizeToRestore += fi.FileMeta.FileSize
-					if fi.FileMeta.Type == mydump.SourceTypeParquet {
-						numberRows, err := mydump.ReadParquetFileRowCountByFile(ctx, rc.store, fi.FileMeta)
-						if err != nil {
-							return errors.Trace(err)
-						}
-						if m, ok := metric.FromContext(ctx); ok {
-							m.RowsCounter.WithLabelValues(metric.StateTotalRestore, tableName).Add(float64(numberRows))
```
Comment (D3Hunter) [inline lightning/pkg/importer/import.go:1499]: we also init the metric here,

**Review by D3Hunter — APPROVED** [review]
still too slow to read all parquet files to get row count in serial, we can optimize it in later pr

**Pull request merged**